### ISSUE TITLE
Missing wal segments on archive-push  take 2

### ISSUE THREAD

**Issue by mmontagna** [pgbackrest/pgbackrest#1433]
I am a friend and colleague of Lev's and I'm following up about the archive push failures mentioned here: https://github.com/pgbackrest/pgbackrest/issues/1116 

Opening a new ticket for clarity as some config has changed and the old ticket reported two distinct issues.

1. pgBackRest version: 2.32

2. PostgreSQL version: 12.6

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
Ubuntu 18.04

4. Did you install pgBackRest from source or from a package?
A .deb package from the repository

5. Please attach the following as applicable:
pgbackrest.conf

```
[main]
pg1-path=/var/lib/postgresql/12/main

[global]
process-max=8
io-timeout=600
repo1-retention-diff=2
repo1-retention-full=2
repo1-type=s3
repo1-path=/blue_arm_node_3
repo1-s3-bucket=mybucket
repo1-s3-endpoint=s3.amazonaws.com
repo1-s3-region=us-east-1
spool-path=/var/spool/pgbackrest
archive-async=y
log-subprocess=y
log-level-file=info
compress-type=lz4

[global:archive-push]
compress-level=1
process-max=32
archive-push-queue-max=1TB

[global:backup]
process-max=12
compress-level=3
db-timeout=1800
protocol-timeout=1830
archive-timeout=3600
start-fast=y
expire-auto=n

[global:archive-get]
process-max=4
```

7. Describe the issue: 
We are regularly encountering an issue where archiving drops a wal segment yet continues to archive later segments.

I've been doing some digging and in all the cases I've seen it looks like pgbackrest successfully uploads the missing wal segment parts to s3, but then fails to finish the multi-part upload. I can find the missing segments via the s3 list-multipart-uploads api. 

Calling the s3 list-parts api results in a series of part uploads which are always a series of parts with size 5242880 followed by a single part with a size less than 5242880. I am able to finalize the multi-part upload with the complete-multipart-upload and the segments are complete (replicas consume/replay them without issue). 

All of this makes me think there is a bug/or potential issue in the code which finalizes the s3 multi-part upload but I'm not familiar enough with C to know offhand what it might be. 

Hopefully this is at least somewhat helpful. I'll continue to poke at the write/close code. Please let me know if you'd like anything else from us.

Best,
Marco

**Comment by mmontagna** [comment]
Also side note but I'm surprised you need to use multi-part uploads for standard sized wal segments.

**Comment by mmontagna** [comment]
I think it's not sufficient to check http response codes for finishing multi-part uploads.  See

https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html
```
Processing of a Complete Multipart Upload request could take several minutes to complete. After Amazon S3 begins processing the request, it sends an HTTP response header that specifies a 200 OK response. While processing is in progress, Amazon S3 periodically sends white space characters to keep the connection from timing out. Because a request could fail after the initial 200 OK response has been sent, it is important that you check the response body to determine whether the request succeeded.
```

My very crude reading of the codebase leads me to believe this might be it ^, https://github.com/pgbackrest/pgbackrest/blob/aed3d468a1557474eba4a24854732fc09581b265/src/storage/s3/storage.c#L446 but I'm not sure yet.

**Comment by sfrost** [comment]
Wow is that an insanely horrible API.

I'll also note that the same documentation you link to also has this: "If the action is successful, the service sends back an HTTP 200 response."

I'm suspicious that they realized at some point later that they sometimes return 200 w/ an error code in the body but instead of *fixing* that decided that they'd just document this instead, which is also terrible.

One has to wonder what the various s3 gateways and such do too.  This is a nasty rabbit hole to have to go down.  They also don't actually document the Error result as being a possible Response Element, all they do is provide an example, which is also horrible.

Basically, this is terrible and broken of Amazon/s3, but ultimately something we're going to have to deal with somehow.

While we should probably check the request body on response, I'm wondering if maybe we should also do something like actually check that the file ends up getting saved after by asking for it..  Unfortunately, that can take a while and the process will just have to sit there and poll for it if we decided to go down that route.  Still, seems like it might be worth it, to give us a bit more comfort that the file actually got stored correctly.

**Comment by dwsteele** [comment]
Yeah -- hard to prove it but I feel like they have changed something here in the last few years.

I think the best thing to start with is to check the response and at least the size, like we do for Azure and GCS. We've been planning on that, but now that we have some confirmation of what is going on here (thanks @mmontagna!) we'll bump up the priority and get it in the next release. It's a fairly minor change.

Polling for the file afterwards is something we used to do but it ended up being really slow on some S3 clones.

**Comment by sfrost** [comment]
Checking the size sounds like a generally good idea- but how are you going to do that?

I agree with actually getting the response body and checking it to make sure that it's a "CompleteMultipartUploadResult" and not an "Error" in the body.  Note that, as mentioned, AWS will continue to feed us whitespace while it is processing the upload.  Hopefully won't be too hard to deal with though.

Perhaps we should make it optional to poll for the file afterwards, maybe even make it the default if we know that we're talking to an s3 gateway that is typically fast, but off otherwise?

Do we anticipate this resulting in a lot more time being taken for these uploads to be completed?  That is- have we always been waiting for the other end to properly close the connection anyway and just ignoring the result body, or are we actively closing the connection once we get the 200, meaning that changing this will end up with all uploads taking longer as we'll have to wait for the actual response?  Mostly wondering if we'll want to suggest to people to increase their process-max if they push a lot of WAL to s3 after we implement this change.

**Comment by dwsteele** [comment]
> That is- have we always been waiting for the other end to properly close the connection anyway and just ignoring the result body, or are we actively closing the connection once we get the 200.

We always read the entire response since we reuse connections and we need to make sure we are ready for the next request. We are not parsing or checking the result, though, since 200 generally means success on its own.

**Comment by sfrost** [comment]
ah, ok, so we just need to check the body then and that shouldn't add much time at all, ok.

**Comment by mmontagna** [comment]
Thanks for the fast response! I am willing to bet this is a very rare error condition for most systems and is only manifesting on ours because we're pushing TB of wal an hour to a single s3 bucket (causing frequent rate limit errors).

**Comment by sfrost** [comment]
That sounds rather fun...  pgbackrest is generally working well for you with that much WAL?  Would love to chat with you further about what you're doing with pgbackrest, maybe do some kind of case study type of thing. :)

**Comment by mmontagna** [comment]
Yeah it's certainly fun...We're definitely pushing postgres/pgbackrest pretty hard here. I just checked and it's like 0.5TB of wal an hour per node at peak and we've got quite a few nodes. 

Happy to chat further. I'll shoot you an email.

**Comment by dwsteele** [comment]
@mmontagna PR #1438 addresses this issue. I have not tested it against AWS S3 yet but it works fine on Minio. Have a look and let me know what you think.

Is this something you'd be able/willing to test? The patch to ` src/storage/s3/write.c` is pretty straightforward.

**Comment by mmontagna** [comment]
Yep! We can test this out next week. Thanks

**Comment by dwsteele** [comment]
I have now tested this on AWS and it appears to work fine.

I'm wondering if it is worth attempting a retry of the multi-part completion when it fails. The entire upload will be retried automatically, but the question is whether it would be beneficial to retry just the completion. My feeling is no, and it would definitely require a bigger patch because we'd need to have a callback in the HTTP request to allow us to error on criteria other than the HTTP result code, which feels pretty grotty. Throwing an error that starts with `HTTP request failed with 200 (OK)` is already awful.

**Comment by mmontagna** [comment]
I think this is sufficient, at least in the case of wal retrying the entire upload should be cheap. 

It may be worse for backup files as they can be larger and retries more expensive, but still I think this is rare enough that it won't matter. 

We have tested the patch and verified it doesn't appear to break anything. We're going to try a slow rollout and will monitor whether this impacts backup times at all. I don't think we'll know for sure whether it fixes the missing WAL until we've gone several weeks without an issue, but this looks very promising so far!

**Comment by dwsteele** [comment]
Excellent, thanks for trying it out. Our next release is August 9th so we have plenty of time to see how it goes.

**Comment by dwsteele** [comment]
We are going to review this patch and commit it, so closing this issue. Let us know if the fix isn't working.

**Comment by mmontagna** [comment]
Just following up we've been testing this patch, we really put it through it's paces with archiving, and it seems to be holding up well. No missing segments since the patch.

**Comment by dwsteele** [comment]
Excellent. Thanks for the update!
